src/mon/OSDMonitor.cc

```diff
@@ -4425,6 +4438,31 @@ void OSDMonitor::tick()
 	  }
 	}
 
+	if (g_conf().get_val<string>("mon_custom_subtree_down_out_type").length()) {
+	  int type = osdmap.crush->get_type_id(g_conf().get_val<string>("mon_custom_subtree_down_out_type"));
+	  if (type > 0) {
+	    if (osdmap.containing_subtree_is_down(g_ceph_context, o, type, &down_cache)) {
+	      int parent_id = osdmap.crush->get_parent_of_type(o, type);
+	      map<int,utime_t>::iterator i = down_pending_out_subtree.find(parent_id);
+	      if (i != down_pending_out_subtree.end()) {
+		utime_t down_time = now - i->second;
+		utime_t cut(g_conf().get_val<uint64_t>("mon_custom_subtree_down_out_interval"), 0);
+		if (down_time <= cut) {
+		  down_pending_out[o] = now;
+		  continue;
+		} else {
+		  // the subtree should down
+		  custom_subtree_marked_out.insert(parent_id);
+		  force_down = true;
```

Comment:
I'm confused; why do we have to add a separate ```force_down``` flag when all we're doing in this loop is supposed to *prevent* the early out'ing of a full host?
(Also, why is it ```force_down``` instead of ```force_out```?

Reply:
force_out is fine. I want use the flag to bypass the check of osd grace interval when satisfy the subtree down out interval. otherwise, the subtree timestamp  will fresh, this is the case I want  avoid.
any suggestion? thanks!

Reply:
So with the flag, osds of full host will down out iff satisfy the subtree down out interval.